### ISSUE TITLE
@RequestParam annotation problem in POST methods on services.

### ISSUE THREAD

**Issue by ismail-codar** [advantageous/qbit#702]
I am working on creating typescript client and qbit server codes from swagger schema. Currently close to the end. But I found some problems while testing.

Following code  is technically correct (may be) but not running.

``` java
@RequestMapping(value = "/pets")
public class PetsService {
    petstore.server.service.impl.PetsServiceImpl serviceImpl = new petstore.server.service.impl.PetsServiceImpl();

    @RequestMapping(value = "",  //MISSING ??????
              method = RequestMethod.POST, 
              description = "", 
              returnDescription = "405: Invalid input", 
              summary = "Add a new pet to the store", 
             code = -1 // (-1 is not good) HTTP response code -1?????? not good
             contentType = "application/json")
    public void addPet(final Callback<Void> callback, 
        @RequestParam(value = "body", 
                                         required = false, 
                                        description = "Pet object that needs to be added to the store") 
                                 Pet body) {
        serviceImpl.addPet(callback, body);
    }
```

But following code is running. (Only @RequestParam(value = "body"... is removed)

``` java
@RequestMapping(value = "/pets")
public class PetsService {
    petstore.server.service.impl.PetsServiceImpl serviceImpl = new petstore.server.service.impl.PetsServiceImpl();

    @RequestMapping(value = "", method = RequestMethod.POST, description = "", returnDescription = "405: Invalid input", summary = "Add a new pet to the store", code = -1, contentType = "application/json")
    public void addPet(final Callback<Void> callback, 
        Pet body) {
        serviceImpl.addPet(callback, body);
    }
```

It seems to be problems associated with @RequestMapping annotation.

**Comment by ismail-codar** [comment]
https://github.com/advantageous/qbit/blob/master/qbit/core/src/main/java/io/advantageous/qbit/meta/transformer/StandardRequestTransformer.java#L155 

It looks in the case REQUEST section request body must be handled.

**Comment by RichardHightower** [comment]
You are missing some stuff in your `@RequestMapping`.

Fix those and see if it runs.

I am glad you are working on codegen. 

Keep me in the loop.

See my comments in your original post.

**Comment by ismail-codar** [comment]
Here is the some results.
![image](https://cloud.githubusercontent.com/assets/282104/14966929/65b84d5c-10bd-11e6-9d58-3259708d2caa.png)

Here is the Spring MVC target generated code. They are uses value=""... too.
```java
@Controller
@RequestMapping(value = "/pets", produces = { APPLICATION_JSON_VALUE })
@Api(value = "/pets", description = "the pets API")
@javax.annotation.Generated(value = "class io.swagger.codegen.languages.SpringMVCServerCodegen", date = "2016-04-22T13:54:03.943Z")
public class PetsApi {
	@ApiOperation(value = "Add a new pet to the store", notes = "", response = Void.class, authorizations = {
			@Authorization(value = "petstore_auth", scopes = {
					@AuthorizationScope(scope = "write_pets", description = "modify pets in your account"),
					@AuthorizationScope(scope = "read_pets", description = "read your pets") }) })
	@io.swagger.annotations.ApiResponses(value = {
			@io.swagger.annotations.ApiResponse(code = 405, message = "Invalid input") })
	@RequestMapping(value = "", produces = { "application/json", "application/xml" }, consumes = { "application/json",
			"application/xml" }, method = RequestMethod.POST)
	public ResponseEntity<Void> addPet(

			@ApiParam(value = "Pet object that needs to be added to the store") @RequestBody Pet body)
			throws NotFoundException {
		// do some magic!
		return new ResponseEntity<Void>(HttpStatus.OK);
	}
```

**Comment by RichardHightower** [comment]
I will take a look. I know the @RequestMapping work see https://github.com/MammatusTech/qbit-microservices-examples/wiki/QBit-Microservices-Java-Lib-RESTful

Why don't you start with that and then slowing change your code until it is more and more like the working example.

**Comment by RichardHightower** [comment]
https://github.com/MammatusTech/qbit-microservices-examples/tree/master/todo-async

**Comment by RichardHightower** [comment]
There are many working examples. https://github.com/MammatusTech/qbit-microservices-examples

But if you still are having a problem send me whatever stacktraces you find or just check in your code where I can git clone it, and I will take a look.

**Comment by ismail-codar** [comment]
I have sended https://github.com/advantageous/qbit/pull/703. This is only problem simulating test. It is not solution.

Problem is related https://github.com/advantageous/qbit/blob/master/qbit/core/src/main/java/io/advantageous/qbit/meta/transformer/StandardRequestTransformer.java#L155

Because if the POST body parameter has @RequestParam annotation it executes in `case REQUEST:` setion incorrectly otherwise executes `case BODY:` section correctly.

It can be this way. Depending on your choice. I can use without putting @RequestParam annotation on POST body like parameters.

**Comment by RichardHightower** [comment]
Ok.. I think I see the problem now. Let me do some research.

**Comment by RichardHightower** [comment]
You need to use no annotation for the body.

@RequestParam is for query string and form params only.

```java

/**
 * Defines a request parameter used to map service method parameters to HTTP request params.
 *
 * @author Rick Hightower
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(value = {ElementType.PARAMETER})
public @interface RequestParam {
```

This might be different than they way spring does it. 
Now I finally get your comments and struggle. Sorry.. Balancing many spinning plates. Seems that you understand as well.

Notice that the Spring example has a `@RequestBody`. QBit does not have this. The param that is not a callback or `@HeaderParam` or `@RequestParam` is assumed to be the body.

So in short (which I think you figured out already.. good job), `@RequestParam` is the same as Spring's `@RequestParam`, and QBit does not have `@RequestBody` but it should. 

I will open an issue for `@RequestBody` and close this issue if that is ok with you. 

If you want to tackle `@RequestBody`, give it a shot. I think we need it for the meta data (description).

**Comment by RichardHightower** [comment]
Since you are deep in the code. Try adding it and sending a PR. I would love the help.

**Comment by ismail-codar** [comment]
In theory @RequestBody is required but not immediate. Not required to follow the same path with spring.

Currently swagger schema to code generator near close with some incomplete features (autentication etc.).
https://github.com/ismail-codar/codegen-maven/

Example outputs: https://github.com/ismail-codar/codegen-example-petstore-server/blob/master/src/main/java/petstore/server/service/PetsService.java and https://github.com/ismail-codar/codegen-example-petstore-server/blob/master/src/js/clientApi.ts

**Comment by ismail-codar** [comment]
Another topic: Swagger schema to service generation and after this swagger schema generation form services with AdminBuilder must be same results.
I will send some PR's for swagger compatibility.

**Comment by RichardHightower** [comment]
ok.. sounds good.

On Tue, May 3, 2016 at 7:22 PM, ismail-codar <notifications@github.com>
wrote:

> Another topic: Swagger schema to service generation and after this swagger
> schema generation form services with AdminBuilder must be same results.
> I will send some PR's for swagger compatibility.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/advantageous/qbit/issues/702#issuecomment-216725287>
>



-- 
*Rick Hightower*
(415) 968-9037
Profile <http://www.google.com/profiles/RichardHightower>

**Comment by RichardHightower** [comment]
Good stuff! Thanks.

On Tue, May 3, 2016 at 6:52 PM, ismail-codar <notifications@github.com>
wrote:

> In theory @RequestBody <https://github.com/RequestBody> is required but
> not immediate. Not required to follow the same path with spring.
>
> Currently swagger schema to code generator near close with some incomplete
> features (autentication etc.).
> https://github.com/ismail-codar/codegen-maven/
>
> Example outputs:
> https://github.com/ismail-codar/codegen-example-petstore-server/blob/master/src/main/java/petstore/server/service/PetsService.java
> and
> https://github.com/ismail-codar/codegen-example-petstore-server/blob/master/src/js/clientApi.ts
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/advantageous/qbit/issues/702#issuecomment-216719135>
>



-- 
*Rick Hightower*
(415) 968-9037
Profile <http://www.google.com/profiles/RichardHightower>
